sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala

```diff
@@ -209,13 +211,16 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
 
           val updatedTable = inferIfNeeded(relation, options, fileFormat, Option(fileIndex))
 
+          // Spark SQL's data source table now support static and dynamic partition insert. Source
+          // table converted from Hive table should always use dynamic.
+          val enableDynamicPartition = options.updated("partitionOverwriteMode", "dynamic")
```

Comment:
I'm not sure about this part. Why it should always use dynamic?

Reply:
Please refer https://issues.apache.org/jira/browse/SPARK-20236.
TL;DR version: dynamic follows hive's style, static is the spark style which is different with hive. When converting InsertIntoHiveTable, we definitely want the hive behaviour. So it takes preference over `spark.sql.sources.partitionOverwriteMode`